Nest joins only if required and fix some issues with complex joins

katzyn opened the pull request:
1. I found a complex query that causes NPE in both development and latest released (1.4.196) versions. If joined columns are not indexed query planner may not set indexes for some deeply nested table filters. It looks like this situation is expected, because `TableFilter.setPlanItem()` have such checks, but query cannot be executed later if `TableFilter` does not have an index. So I add a method that sets missing indexes to scan indexes. This situation was not covered by tests. New test covers it, and I compared results with other databases. But this query contains only '=' comparison operators, so output is very simple.

2. Database name in `TestOuterJoins` was incorrect, so failures in `TestNestedJoins` may cause unrelated failures in `TestOuterJoins` too because some tests use a shared database, probably due to copy-and-paste errors.

3. NPE in `IndexCondition.toString()` is fixed. I found it with debugger.

4. Nesting of joins now checks if this nesting is actually needed. This reduces complexity of trees with table filters on some queries.

katzyn committed:
Fix database name in TestOuterJoins

katzyn committed:
Fix NPE in IndexCondition.toString()

katzyn committed:
Fix NPE with some complex join conditions

katzyn committed:
Nest joins only if required

katzyn commented:
It fails on Travis with assertion errors, my `TestAll` configuration didn't have `-ea` flag, so I missed that. I'll take a look on it.

grandinj commented on h2/src/main/org/h2/table/TableFilter.java:
```diff
@@ -265,15 +265,38 @@ public void setPlanItem(PlanItem item) {
+     *
+     * @param table filter to start from
+     */
+    private void setScanIndexes(TableFilter filter) {
```
seems a little odd to have this method take a param instead of being an instance method of TableFilter?

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1799,26 +1796,25 @@ private TableFilter readJoin(TableFilter top, boolean nested) {
-                null);
-        top.setNestedJoin(n);
-        return top;
+    private void addJoin(TableFilter top, TableFilter join, boolean outer, Expression on) {
```
please add comment to this method "nest joins only if required"

katzyn commented:
This code causes assertion failures:
https://github.com/h2database/h2database/blob/7e5ab118663a856c10cd60ec4ea073c8674d233e/h2/src/main/org/h2/table/TableView.java#L471-L473

Failed assertion:
https://github.com/h2database/h2database/blob/7e5ab118663a856c10cd60ec4ea073c8674d233e/h2/src/main/org/h2/index/ViewIndex.java#L169-L171

I used this method to assign at least some index if there are no better solution from the query planner (because there are no indexes or may be due to other issues, it's not easy to find a query with incomplete query plan). Looks like this assertion is not valid, but I'll try to add a some test with views and joins to check that query will be executed properly.

katzyn committed:
Remove unexpected assertion, null is valid here

katzyn committed:
Use this instance instead of parameter in TableFilter.setScanIndexes()

katzyn committed:
Add javadoc to Parser.addJoin()

katzyn committed:
Use field directly

katzyn commented:
I didn't find a reason for such assertion, this parameter is not required to have any useful value in called code. So I removed it. I added one more simple test that triggered such assertion.
